internal/prometheusbp/spectest/spec.go

```diff
@@ -0,0 +1,203 @@
+package spectest
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+
+	"github.com/reddit/baseplate.go/errorsbp"
+)
+
+const (
+	client = "client"
+	server = "server"
+)
+
+var (
+	errPrefix         = errors.New("the prefix is not at the beginning of the metric name")
+	errClientServer   = fmt.Errorf("missing %q or %q as the second part of the metric name", client, server)
+	errLength         = errors.New("metric name should have a minimum of three parts")
+	errNotFound       = errors.New("metric not found")
+	errLabelNotFound  = errors.New("label not found")
+	errPrometheusLint = errors.New("prometheus GatherAndLint")
+)
+
+// ValidateSpec validates that the Prometheus metrics being exposed from baseplate.go
+// conform to the baseplate spec.
+func ValidateSpec(t *testing.T, metricPrefix string, wantMetricCount int) {
+	t.Helper()
+
+	var batch errorsbp.Batch
+	batch.AddPrefix("validate spec", validateSpec(metricPrefix))
+
+	if len(batch.GetErrors()) > 0 {
+		t.Errorf(batch.Compile().Error())
+	}
+}
+
+func validateSpec(prefix string) errorsbp.Batch {
+	var batch errorsbp.Batch
+
+	metricFam, err := prometheus.DefaultGatherer.Gather()
+	if err != nil {
+		return batch
+	}
+
+	allMetricNames := buildMetricNames(prefix)
+
+	for _, m := range metricFam {
+		name := *m.Name
+
```

Comment:
My slight concern with the construction of the loop (i.e. going through all metrics) is that we might accidentally trip over some additional metrics.  For example, if the grpc library does something like `grpc_registered_proto_service_count` (a possibly-bad example of something that is common between the client and server) then it would fail the `validatePromLint` even though the library could still be spec compliant.

I would probably
* use a loop to construct the desired metric names
* use a loop over the gathered metrics to collect a map of metric name to definition info
* loop through desired names and raise an error if they aren't found in the actual map, and lint and check labels for the ones that are.

Reply:
this method `buildMetricNames(prefix)` creates a list of desired metrics names so that your concern of `grpc_registered_proto_service_count` getting tripped on shouldn't occur. `validatePromLint` will only be executed if `buildMetricNames(prefix)` returns the metric name.  

Down on line 80 it errors if a metric is not found.

Reply:
indeed 😅